collect_app/src/main/java/org/odk/collect/android/utilities/QRCodeUtils.java

```diff
@@ -91,50 +84,88 @@ private static BinaryBitmap getBinaryBitmap(Bitmap bitmap) {
+        return Observable.create(emitter -> {
+            String preferencesString = SharedPreferencesUtils.getJSONFromPreferences(selectedPasswordKeys);
+
+            MessageDigest md = MessageDigest.getInstance("MD5");
+            md.update(preferencesString.getBytes());
+            byte[] messageDigest = md.digest();
+
+            boolean shouldWriteToDisk = true;
+            Bitmap bitmap = null;
+
+            // check if settings directory exists, if not then create one
+            File writeDir = new File(Collect.SETTINGS);
+            if (!writeDir.exists()) {
+                if (!writeDir.mkdirs()) {
+                    Timber.e("Error creating directory " + writeDir.getAbsolutePath());
+                }
+            }
+
+            File mdCacheFile = new File(MD5_CACHE_PATH);
+            if (mdCacheFile.exists()) {
+                byte[] cachedMessageDigest = FileUtils.read(mdCacheFile);
+
+                /*
+                 * If the messageDigest generated from the preferences is equal to cachedMessageDigest
+                 * then don't generate QRCode and read the one saved in disk
+                 */
+                if (Arrays.equals(messageDigest, cachedMessageDigest)) {
+                    Timber.i("Loading QRCode from the disk...");
+                    bitmap = BitmapFactory.decodeFile(QR_CODE_FILEPATH);
+                    shouldWriteToDisk = false;
+                }
             }
 
-            Map<EncodeHintType, ErrorCorrectionLevel> hints = new HashMap<>();
-            hints.put(EncodeHintType.ERROR_CORRECTION, ErrorCorrectionLevel.L);
-            QRCodeWriter qrCodeWriter = new QRCodeWriter();
-            BitMatrix bitMatrix = qrCodeWriter
-                    .encode(compressedData, BarcodeFormat.QR_CODE, 400, 400, hints);
+            // If the file is not found in the disk or md5Hash not matched
+            if (bitmap == null) {
+                Timber.i("Generating QRCode...");
+                bitmap = generateQRBitMap(preferencesString, QR_CODE_SIDE_LENGTH);
+                shouldWriteToDisk = true;
+            }
 
-            int width = bitMatrix.getWidth();
-            int height = bitMatrix.getHeight();
+            if (bitmap != null) {
+                // Send the QRCode to the observer
+                emitter.onNext(bitmap);
 
-            Bitmap bmp = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565);
+                // Save the QRCode to disk
+                if (shouldWriteToDisk) {
+                    Timber.i("Saving QR Code to disk... : " + QR_CODE_FILEPATH);
+                    FileUtils.saveBitmapToFile(bitmap, QR_CODE_FILEPATH);
 
-            for (int x = 0; x < width; x++) {
-                for (int y = 0; y < height; y++) {
-                    bmp.setPixel(x, y, bitMatrix.get(x, y) ? Color.BLACK : Color.WHITE);
+                    // update .md5 file
+                    Timber.i("Updated .md5 file contents");
```

Comment:
When you rename the .md5 file, how about using some constant part of it here, so the file name will always match?